packages/react-flicking/src/react-flicking/Flicking.tsx

```diff
@@ -1,188 +1,225 @@
+        renderExternal: true
+      },
     ).on({
-      moveStart: (e: FlickingEvent) => this.props.onMoveStart!(e),
-      move: (e: FlickingEvent) => this.props.onMove!(e),
-      moveEnd: (e: FlickingEvent) => this.props.onMoveEnd!(e),
-      holdStart: (e: FlickingEvent) => this.props.onHoldStart!(e),
-      holdEnd: (e: FlickingEvent) => this.props.onHoldEnd!(e),
-      select: (e: FlickingEvent) => this.props.onSelect!(e),
-      needPanel: (e: NeedPanelEvent) => this.props.onNeedPanel!(e),
-      change: (e: FlickingEvent) => this.props.onChange!(e),
-      restore: (e: FlickingEvent) => this.props.onRestore!(e),
-      visibleChange: (e: VisibleChangeEvent) => {
-        this.props.onVisibleChange!(e);
-        this.forceUpdate();
+      [EVENTS.READY]: (e: ReadyEvent) => props.onReady(e),
+      [EVENTS.BEFORE_RESIZE]: (e: BeforeResizeEvent) => props.onBeforeResize(e),
+      [EVENTS.AFTER_RESIZE]: (e: AfterResizeEvent) => props.onAfterResize(e),
+      [EVENTS.HOLD_START]: (e: HoldStartEvent) => props.onHoldStart(e),
+      [EVENTS.HOLD_END]: (e: HoldEndEvent) => props.onHoldEnd(e),
+      [EVENTS.MOVE_START]: (e: MoveStartEvent) => props.onMoveStart(e),
+      [EVENTS.MOVE]: (e: MoveEvent) => props.onMove(e),
+      [EVENTS.MOVE_END]: (e: MoveEndEvent) => props.onMoveEnd(e),
+      [EVENTS.CHANGE]: (e: ChangeEvent) => props.onChange(e),
+      [EVENTS.RESTORE]: (e: RestoreEvent) => props.onRestore(e),
+      [EVENTS.SELECT]: (e: SelectEvent) => props.onSelect(e),
+      [EVENTS.NEED_PANEL]: (e: NeedPanelEvent) => props.onNeedPanel(e),
+      [EVENTS.VISIBLE_CHANGE]: (e: VisibleChangeEvent) => {
+        props.onVisibleChange!(e);
+        if (flicking.renderOnlyVisible) {
+          this.setState({});
+        }
       },
+      [EVENTS.REACH_EDGE]: (e: ReachEdgeEvent) => props.onReachEdge(e),
     });
-    const children = this.getChildren();
-    this.jsxDiffer = new ListDiffer<string>(children.map(child => `${child.key}`));
 
-    if (this.props.status) {
-      this.setStatus(this.props.status);
+    if (props.circular) {
+      flicking.renderer.elementManipulator.on("orderChanged", e => {
+        this.setState({});
+      });
     }
-    this.checkPlugins();
-    this.checkCloneCount();
 
-    if (this.props.renderOnlyVisible) {
-      this.forceUpdate();
-    }
+    this._nativeFlicking = flicking;
+
+    const children = this._getChildren(this.props.children);
+    this._jsxDiffer = new ListDiffer(children, panel => panel.key!);
+    this._pluginsDiffer = new ListDiffer<any>();
+
+    this._checkPlugins();
+
+    this.setState({});
   }
 
   public componentWillUnmount() {
-    this.destroy({ preserveUI: true });
+    this._nativeFlicking.destroy();
   }
 
-  // private
-  private checkPlugins() {
-    const { list, added, removed, prevList } = this.pluginsDiffer.update(this.props.plugins!);
+  public shouldComponentUpdate(nextProps: this["props"]) {
+    const children = this._getChildren(nextProps.children);
```

Comment:
The code of shouldComponentUpdate and componentDidUpdate is too long. Is there a reason to handle it in the framework?

Reply:
Well my code was not sanitized so I've just updated them to use the same code.